pkg/knowledge_base2/reader/properties.go

```diff
@@ -0,0 +1,227 @@
+		property.Properties[name] = child
+	}
+}
+
+func (p Properties) Clone() Properties {
+	newProps := make(Properties, len(p))
+	for k, v := range p {
+		newProps[k] = v.Clone()
+	}
+	return newProps
+}
+
+func (p *Property) Clone() *Property {
+	cloned := *p
+	cloned.Properties = make(Properties, len(p.Properties))
+	for k, v := range p.Properties {
+		cloned.Properties[k] = v.Clone()
+	}
+	return &cloned
+}
+
+func InitializeProperty(ptype string) (knowledgebase.Property, error) {
+	if ptype == "" {
+		return nil, fmt.Errorf("property does not have a type")
+	}
+	parts := strings.Split(ptype, "(")
+	p, found := initializePropertyFunc[parts[0]]
+	if !found {
+		return nil, fmt.Errorf("unknown property type '%s'", ptype)
+	}
+	var val string
+	if len(parts) > 1 {
+		val = strings.TrimSuffix(strings.Join(parts[1:], "("), ")")
+	}
+	return p(val)
+}
+
+var initializePropertyFunc = map[string]func(val string) (knowledgebase.Property, error){
+	"string": func(val string) (knowledgebase.Property, error) { return &properties.StringProperty{}, nil },
+	"int":    func(val string) (knowledgebase.Property, error) { return &properties.IntProperty{}, nil },
+	"float":  func(val string) (knowledgebase.Property, error) { return &properties.FloatProperty{}, nil },
+	"bool":   func(val string) (knowledgebase.Property, error) { return &properties.BoolProperty{}, nil },
+	"resource": func(val string) (knowledgebase.Property, error) {
+		id := construct.ResourceId{}
+		err := id.UnmarshalText([]byte(val))
+		if err != nil {
+			return nil, fmt.Errorf("invalid resource id for property type %s: %w", val, err)
+		}
+		return &properties.ResourceProperty{
+			AllowedTypes: construct.ResourceList{id},
+		}, nil
+	},
+	"map": func(val string) (knowledgebase.Property, error) {
+		if val == "" {
+			return &properties.MapProperty{}, nil
+		}
+		args := strings.Split(val, ",")
+		if len(args) != 2 {
+			return nil, fmt.Errorf("invalid number of arguments for map property type")
+		}
+		keyVal, err := getPropertyType(args[0])
+		if err != nil {
+			return nil, err
+		}
+		valProp, err := getPropertyType(args[1])
+		if err != nil {
+			return nil, err
+		}
+		return &properties.MapProperty{KeyProperty: keyVal, ValueProperty: valProp}, nil
+	},
+	"list": func(val string) (knowledgebase.Property, error) {
+		if val == "" {
+			return &properties.ListProperty{}, nil
+		}
+		itemProp, err := getPropertyType(val)
+		if err != nil {
+			return nil, err
+		}
+		return &properties.ListProperty{ItemProperty: itemProp}, nil
+	},
+	"set": func(val string) (knowledgebase.Property, error) {
+		if val == "" {
+			return &properties.SetProperty{}, nil
+		}
+		itemProp, err := getPropertyType(val)
+		if err != nil {
+			return nil, err
+		}
+		return &properties.SetProperty{ItemProperty: itemProp}, nil
+	},
+	"any": func(val string) (knowledgebase.Property, error) { return &properties.AnyProperty{}, nil },
+}
+
+func getPropertyType(ptype string) (knowledgebase.Property, error) {
```

Comment:
This doesn't seem necessary, can just call `InitializeProperty` instead, right?

Reply:
then go complains about a possible endless loop, so this was my workaround. Is there a way for go to like methods calling eachother?

Reply:
You can have recursive functions, or mutually recursive functions in Go. What was the specific error you were getting?